### PR TITLE
Improving the LOD manager behavior for better scaling

### PR DESCRIPTION
**Pull request by samcake**
This PR introduces a different algorithm in the LOD manager in order to get better scaling behavior on the rendering tasks (render thread, present thread and GPU workload) in order to target a constant framerate regardless of the load and complexity of the content in view.

there are several changes in this pr that provide this improvement:
- The rendering LOD is now expressed as the minimum Apparent Angle (in degrees) in the scene, we call it _LODAngle_. 
   Any render item whose bounding box evaluate to an apparent angle smaller than the LODAngle is now culled.
   This replaces (in the client) the previous LOD value and the underneath _octreeSizeScale used as a relative size to the Entity tree octree. The _octreeSizeScale is still used to configure the queries to the server. this doesn't change with the new LODAngle approach, just that the _octreeSizeScale is computed FROM the LODAngle.

- The new LOD regulator algorithm acts as a simple Proportional controller which input the measured Render FPS, compare it against a Target FPS and output the LOD Angle to be used for future frames for culling. The dynamic behavior of the regulator in various environment and on various platforms is the key to the success of this PR. it should scale to any type of content on any device (low or high end). 
  
- The target FPS is defined in LODManager from the "World details setting" available in the Graphics setting dialog. The 3 settings available (high, mid, low) with the default being "mid" are translating into a target FPS to the LODManager. The target FPS is a multiple of the hardware display device ideal frequency in use + a few Hz for security. the mapping is as follow:
 -- low:   60Hz monitor : 90Hz HMD    + 5hz
 -- mid:  30Hz monitor : 45Hz HMD     + 5hz
 -- high: 15Hz monitor : 23Hz HMD     + 5hz
IN this PR we moved the code that was computing the mapping from the world details setting to the target FPS (the LOD DecreaseFPS previously) to the LOD manager for consistency.

- The "render FPS" is evaluated from the measured time of all the threads involved in producing the rendering: Render thread, Present thread or GPU thread. The "render FPS" is the slowest of the 3 threads since it is the rendering bottleneck. The time spent in Present thread per frame is partly spent in the Batch execution and in the displayPlugin::present call that is synchronized with the refresh rate of the hardware display. So when the Batch time is much smaller than the total present Time, and present Time is seen as the bottleneck and limits artificially the LOD quality as a consequence. we are trying to identify this case and react better by using a combination of the both present and batch time instead of just the present time to evaluate the "render FPS".

- Improved a bunch of the qml widgets used to display the LOD related work available from the LOD.js tool in scripts/developer/utilities/render

- Cleaned the default Stats to show less of the gpu specific info about resource sizes, these are visible in expanded mode. Added the Drawcalls number

## TEST PLAN
This PR needs to be tested:
on every platforms with and without HMD
at every "world details settings" available (low, mid high) with the understanding of the target FPS expected to be experienced.
The target FPS set should be the render and present FPS observed from the stats info (pressing the '/' key)

In various domains with different quantities  of content.
In particular in Engine-dev
Running with at least 300 bots in them.

Note that LODManager is not used on android yet, so no changes should be experienced on Android
## TO DO
- tablet is disappearing havey loaded scene, need to be tagged as not culled. probably in a separate PR because it shows only when the LOD Angle is beyond 60 degrees
- ...

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12448/HighFidelity-Beta-PR13891-7dfdc79.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12448/HighFidelity-Beta-Interface-PR13891-7dfdc79.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12454/HighFidelity-Beta-PR13891-0da22db.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12454/HighFidelity-Beta-PR13891-0da22db.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12454/HighFidelity-Beta-Interface-PR13891-0da22db.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12457/HighFidelity-Beta-PR13891-37b4cf6.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12457/HighFidelity-Beta-PR13891-37b4cf6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12457/HighFidelity-Beta-Interface-PR13891-37b4cf6.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12489/HighFidelity-Beta-PR13891-349f782.apk

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12503/HighFidelity-Beta-PR13891-4844f06.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12513/HighFidelity-Beta-PR13891-6c502e7.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12513/HighFidelity-Beta-PR13891-6c502e7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12513/HighFidelity-Beta-Interface-PR13891-6c502e7.exe

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12528/HighFidelity-Beta-PR13891-9e2a540.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12528/HighFidelity-Beta-PR13891-9e2a540.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12528/HighFidelity-Beta-Interface-PR13891-9e2a540.exe

**Comment by qaAdamSmith82** [comment]
Currently Testing (Windows)

**Comment by conklin94122** [comment]
please build this

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12538/HighFidelity-Beta-PR13891-8179de3.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12538/HighFidelity-Beta-PR13891-8179de3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12538/HighFidelity-Beta-Interface-PR13891-8179de3.exe

**Comment by qaAdamSmith82** [comment]
Stopping Testing until review complete

File: scripts/developer/utilities/render/lod.js [hunk 1]
```diff
@@ -54,23 +93,42 @@
...
-        wireEventBridge(onScreen);
-    }
-
-    function fromQml(message) {
+        if (onTablet) {
+            if (url === QMLAPP_URL) {
+                onAppScreen = true;
+            } else { 
+                onAppScreen = false;
+            }
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/render/lod.js]: Not a blocker: I'm wondering why not:  `onAppScreen = (url === QMLAPP_URL);` ?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12550/HighFidelity-Beta-PR13891-1949036.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12552/HighFidelity-Beta-PR13891-dc852bf.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12553/HighFidelity-Beta-PR13891-e0f4c82.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12553/HighFidelity-Beta-PR13891-e0f4c82.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12553/HighFidelity-Beta-Interface-PR13891-e0f4c82.exe

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -6018,7 +6019,7 @@ void Application::updateRenderArgs(float deltaTime) {
                 _viewFrustum.calculate();
             }
             appRenderArgs._renderArgs = RenderArgs(_gpuContext, lodManager->getOctreeSizeScale(),
-                lodManager->getBoundaryLevelAdjust(), RenderArgs::DEFAULT_RENDER_MODE,
+                lodManager->getBoundaryLevelAdjust(), lodManager->getLODAngleHalfTan(), RenderArgs::DEFAULT_RENDER_MODE,
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: the `BoundaryLevelAdjust` part of the LOD and query scheme is really old and probably should be removed... it was the idea of adjusting the Octree level up or down that would normally be implied by the calculated "size scale" option. we have had this hard coded for a while and so we can probably just remove it.

File: interface/src/LODManager.cpp [hunk 3]
```diff
@@ -129,73 +154,123 @@ void LODManager::autoAdjustLOD(float realTimeDelta) {
...
-        case 1: {
-            granularityFeedback = QString(" at half of standard granularity.");
-        } break;
-        case 2: {
-            granularityFeedback = QString(" at a third of standard granularity.");
-        } break;
-        default: {
-            granularityFeedback = QString(" at 1/%1th of standard granularity.").arg(boundaryLevelAdjust + 1);
-        } break;
+    case 0: {
```
Comment (ZappoMan) [inline interface/src/LODManager.cpp]: like all of this can go.... if you remove the boundary level adjust... which is essentially hard coded to 0 right now.

can do in different PR

File: interface/src/Application.cpp (hunk 2 above)
Comment (samcake) [reply interface/src/Application.cpp]: cool , i ll remove that

**Comment by samcake** [comment]
build this please

**Comment by samcake** [comment]
build this please

File: interface/src/LODManager.cpp (hunk 3 above)
Comment (samcake) [reply interface/src/LODManager.cpp]: ok, i ll merge this one first
and yes add another pr to clean up these

File: interface/src/LODManager.cpp [hunk 4]
```diff
@@ -48,79 +36,116 @@ float LODManager::getLODIncreaseFPS() const {
...
+    // Target fps and current fps based on latest measurments
+    float targetFPS = getLODTargetFPS();
+    float currentNowFPS = (float)MSECS_PER_SECOND / _nowRenderTime;
+    float currentSmoothFPS = (float)MSECS_PER_SECOND / _smoothRenderTime;
+ 
+    // Compute the Variance of the FPS signal (FPS - smouthFPS)^2
+    // Also scale it by a percentage for fine tuning (default is 100%)
+    float currentVarianceFPS = (currentSmoothFPS - currentNowFPS);
+    currentVarianceFPS *= currentVarianceFPS;
+    currentVarianceFPS *= _pidCoefs.w;
```
Comment (AndrewMeadows) [inline interface/src/LODManager.cpp]: I think you should be computing a running average of the FPS variance and use that for the comparison logic below rather than the instantaneous "variance" of the most recent measure.  The instantaneous noise will vary (it is noise after all) and the logic should pivot on the characteristic noise amplitude of the signal which must be integrated over multiple measurements.

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12554/HighFidelity-Beta-PR13891-6bd1ab5.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12554/HighFidelity-Beta-PR13891-6bd1ab5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12554/HighFidelity-Beta-Interface-PR13891-6bd1ab5.exe

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12566/HighFidelity-Beta-PR13891-84ad5c7.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12566/HighFidelity-Beta-PR13891-84ad5c7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12566/HighFidelity-Beta-Interface-PR13891-84ad5c7.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12569/HighFidelity-Beta-PR13891-ec18806.apk

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12571/HighFidelity-Beta-PR13891-ec18806.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12571/HighFidelity-Beta-PR13891-ec18806.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12571/HighFidelity-Beta-Interface-PR13891-ec18806.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12571/HighFidelity-Beta-PR13891-ec18806.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12571/HighFidelity-Beta-Interface-PR13891-ec18806.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12581/HighFidelity-Beta-PR13891-5067047.apk

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12586/HighFidelity-Beta-PR13891-5067047.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12581/HighFidelity-Beta-PR13891-5067047.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12581/HighFidelity-Beta-Interface-PR13891-5067047.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12581/HighFidelity-Beta-PR13891-5067047.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12581/HighFidelity-Beta-Interface-PR13891-5067047.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12586/HighFidelity-Beta-PR13891-5067047.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12586/HighFidelity-Beta-Interface-PR13891-5067047.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12586/HighFidelity-Beta-PR13891-5067047.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12586/HighFidelity-Beta-Interface-PR13891-5067047.dmg

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12599/HighFidelity-Beta-PR13891-bd91368.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12599/HighFidelity-Beta-PR13891-bd91368.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12599/HighFidelity-Beta-Interface-PR13891-bd91368.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12599/HighFidelity-Beta-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12599/HighFidelity-Beta-Interface-PR13891-bd91368.exe

**Comment by qaAdamSmith82** [comment]
Currently Testing

**Comment by qaAdamSmith82** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12614/HighFidelity-Beta-PR13891-bd91368.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12614/HighFidelity-Beta-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12614/HighFidelity-Beta-Interface-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12637/HighFidelity-Beta-PR13891-bd91368.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12637/HighFidelity-Beta-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12637/HighFidelity-Beta-Interface-PR13891-bd91368.exe

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12645/HighFidelity-Beta-PR13891-bd91368.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12645/HighFidelity-Beta-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12645/HighFidelity-Beta-Interface-PR13891-bd91368.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12661/HighFidelity-Beta-PR13891-d2dcf6a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12661/HighFidelity-Beta-PR13891-d2dcf6a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12661/HighFidelity-Beta-Interface-PR13891-d2dcf6a.dmg

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12675/HighFidelity-Beta-PR13891-2ced6d9.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12675/HighFidelity-Beta-PR13891-2ced6d9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12675/HighFidelity-Beta-Interface-PR13891-2ced6d9.exe

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
LOD responds well.

**Comment by conklin94122** [comment]
build this please

**Comment by conklin94122** [comment]
@qlimnavarro Please retest when the Windows, Mac, and Android builds all succeed.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12679/HighFidelity-Beta-PR13891-2ced6d9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12679/HighFidelity-Beta-Interface-PR13891-2ced6d9.dmg

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
LOD reacts well to fps changes.
